### PR TITLE
FEATURE: Auto generate and display video preview image

### PR DESCRIPTION
**Pull request by oblakeerickson**
This change will allow auto generated video thumbnails to be used
instead of the black video thumbnail that overlays videos.

Follow up to: 2443446e62bff9dd8a0493d10e5359a5e946a84e

### CODE REVIEW

File: lib/pretty_text.rb [hunk 1]
```diff
@@ -442,6 +443,17 @@ def self.extract_links(html)
...
   end
 
+  def self.add_video_placeholder_image(doc)
+    doc
+      .css(".video-placeholder-container")
+      .each do |video|
+        video_src = video["data-video-src"]
+        video_sha1 = File.basename(video_src, File.extname(video_src))
+        thumbnail = Upload.where("original_filename LIKE ?", "#{video_sha1}.%").first
+        video["data-thumbnail-src"] = "/uploads/default/original/1X/#{thumbnail.sha1}.jpg"
```
Comment (oblakeerickson) [inline lib/pretty_text.rb:453]: Use the extension from the db here, it might not have jpg

File: lib/pretty_text.rb [hunk 2]
```diff
@@ -442,6 +443,21 @@ def self.extract_links(html)
...
+    doc
+      .css(".video-placeholder-container")
+      .each do |video|
+        video_src = video["data-video-src"]
+        video_sha1 = File.basename(video_src, File.extname(video_src))
+        thumbnail = Upload.where("original_filename LIKE ?", "#{video_sha1}.%").last
+        if thumbnail
+          video[
+            "data-thumbnail-src"
+          ] = "/uploads/default/original/1X/#{thumbnail.sha1}.#{thumbnail.extension}"
```
Comment (pmusaraj) [inline lib/pretty_text.rb:456]: Is this multisite and/or subfolder safe? `/uploads/default` might not work on either.

Comment (oblakeerickson) [reply lib/pretty_text.rb:456]: Oh ya, I forgot I hard coded these. Let me see what I can improve.

**Comment by oblakeerickson** [comment]
Also I'm going to remove the current composer preview code for a separate PR so we can ship a better version of that.

**Comment by oblakeerickson** [comment]
We have a system test for video thumbnails we need another one that checks for more than just the topic.image_upload_id, but looks like it currently is disabled

https://github.com/discourse/discourse/blob/1d6455d8a2468fe28ce93699b1f839f666ae7b84/spec/system/composer_uploads_spec.rb#L54

File: app/models/post.rb [hunk 3]
```diff
@@ -1074,7 +1076,7 @@ def link_post_uploads(fragments: nil)
 
     UploadReference.transaction do
       UploadReference.where(target: self).delete_all
-      UploadReference.insert_all(upload_references) if upload_references.size > 0
+      result = UploadReference.insert_all(upload_references) if upload_references.size > 0
```
Comment (pmusaraj) [inline app/models/post.rb:1079]: We're not using `result` further here, are we?

File: lib/pretty_text.rb (hunk 2 above)
Comment (oblakeerickson) [reply lib/pretty_text.rb]: I'm not totally sure the correct way to go about doing this, but I switched to using the upload.url and using the UrlHelper to generate the url like we are doing in other places.

**Review by pmusaraj — APPROVED** [review]
This feels like a good candidate for a "happy path" system test. 

Something like: 

- login as user
- create post with a video upload
- check that rendered post includes a background image on the placeholder video tag

**Comment by pmusaraj** [comment]
> We have a system test for video thumbnails and we can add another one that checks for more than just the topic.image_upload_id, but looks like it currently is disabled
> 
> https://github.com/discourse/discourse/blob/1d6455d8a2468fe28ce93699b1f839f666ae7b84/spec/system/composer_uploads_spec.rb#L54

Ah, just saw this. Maybe we can repurpose that test to check for the background image?

**Comment by oblakeerickson** [comment]
> Ah, just saw this. Maybe we can repurpose that test to check for the background image?

Ya, trying to see if I can get it work for that.

**Comment by oblakeerickson** [comment]
Ooo I got a system test to generate a thumbnail image from an mp4, I'll commit it tomorrow morning.

File: app/models/post.rb (hunk 3 above)
Comment (oblakeerickson) [reply app/models/post.rb:1079]: Nope I must of added it while puts debugging.

**Pull request merged**